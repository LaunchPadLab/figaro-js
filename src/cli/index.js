#!/usr/bin/env node

import { Command } from 'commander'
import registerHerokuCommands from './heroku'

/**
 * This module exposes a CLI that can be run with `yarn figaro`.
 * 
 * Using the `heroku:set` command, you can set values from your configuration file all at once:
 * 
 * `$ yarn figaro heroku:set -e production`
 * 
 * For more information:
 * 
 * `$ yarn figaro -h`
 *  
 * @name CLI
**/

function main (args=process.argv) {
  const program = new Command()
  registerHerokuCommands(program)
  return program.parse(args)
}

// Exported for testing
export default main

if (!module.parent) main()