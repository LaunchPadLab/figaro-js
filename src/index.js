export load from './load'
export read from './read'
export requireKeys from './requireKeys'
