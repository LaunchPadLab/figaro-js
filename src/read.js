import path from 'path'
import { 
  omitObjectValues,
  readYamlFile,
  toStringValues,
} from './utils'

/**
 * A function that loads env variables from a file and returns an object containing them.
 * The file may contain plain key-value pairs ("shared" variables) or key-values pairs nested under environment keys.
 * The latter will only be loaded if the current environment matches the environment key.
 * 
 * **Options**
 * 
 * `read` may be passed an options object containing the following keys:
 * - `path`: The path to the config file (default: './config/application.yml')
 * - `environment`: The string name of the current environment (default: 'development')
 *
 * @name read
 * @type Function
 * @param {object} options - Options for the function as specified above.
 *
 * @example
 * 
 * // Let's say we've got an application.yml that looks like:
 * // SOME_VAR: 'FOO'
 * // development:
 * //   IS_DEV: true
 * // production:
 * //   IS_PROD: true
 *
 * const Figaro = require('figaro-js')
 * const path = require('path')
 * 
 * const myEnv = Figaro.read({
 *    path: path.resolve(__dirname, './application.yml')
 * })
 * 
 * console.log(myEnv) 
 * // -> {
 * //   SOME_VAR: 'FOO',
 * //   IS_DEV: true
 * // }
 *
**/

function read ({
  path:filePath=path.resolve('config/application.yml'),
  environment=process.env.NODE_ENV || 'development',
}={}) {
  // Add vars from application.yml
  const allVars = readYamlFile(filePath)
  // Get shared vars
  const sharedVars = omitObjectValues(allVars)
  // Get environment vars
  const environmentVars = allVars[environment] || {}
  return toStringValues({ ...sharedVars, ...environmentVars })
}

export default read