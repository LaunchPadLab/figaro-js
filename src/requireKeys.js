
/**
 * A function that checks to make sure certain variables have been loaded into `process.env`.
 * If they exist in `process.env`, it returns `true`; if not, it throws an exception.
 * 
 * @name requireKeys
 * @type Function
 * @param {...string} keys - The required keys
 *
 * @example
 * 
 * if (requireKeys('MY_KEY', 'ANOTHER_KEY')) {
 *    runProgram()
 * }
 * 
 * // Above will throw if 'MY_KEY', 'ANOTHER_KEY' are not keys in process.env.
 * 
**/

function requireKeys (...keys) {
  const envKeys = Object.keys(process.env)
  keys.forEach(key => {
    if (!envKeys.includes(key)) throw new Error(`Figaro: Required key "${ key }" not found.`)
  })
  return true
}

export default requireKeys
