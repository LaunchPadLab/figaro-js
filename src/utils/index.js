export findRoot from 'find-root'

// Local

export omitObjectValues from './omitObjectValues'
export readYamlFile from './readYamlFile'