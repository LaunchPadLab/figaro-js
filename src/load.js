import read from './read'

/**
 * A function that loads env variables into `process.env`. 
 * Internally calls {@link read} and merges the result into `process.env`.
 * 
 * **Options**
 * 
 * `load` may be passed an options object containing the following keys:
 * - `path`: The path to the config file (default: './config/application.yml')
 * - `environment`: The string name of the current environment (default: 'development')
 *
 * @name load
 * @type Function
 * @param {object} options - Options for the function as specified above.
 *
 * @example
 * 
 * // Let's say we've got an application.yml that looks like:
 * // SOME_VAR: 'FOO'
 *
 * const Figaro = require('figaro-js')
 * const path = require('path')
 * 
 * console.log(process.env.SOME_VAR) // -> undefined
 * 
 * Figaro.load({
 *  path: path.resolve(__dirname, './application.yml')
 * })
 * 
 * console.log(process.env.SOME_VAR) // -> 'FOO'
 *
**/

function load (options) {
  const readEnv = read(options)
  process.env = { ...process.env, ...readEnv }
}

export default load
