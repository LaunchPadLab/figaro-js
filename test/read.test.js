import { read } from '../src'
import { 
  DEVELOPMENT_ENV,
  PRODUCTION_ENV,
  CONFIG_FILE_PATH,
  INVALID_CONFIG_FILE_PATH,
  JSON_CONFIG_FILE_PATH,
} from './fixtures'

// Global console.warn mock
const warn = jest.spyOn(console, 'warn')
const last = arr => arr[arr.length - 1]
const getLastCall = fn => last(fn.mock.calls)

test('path defaults to {project_root}/config/application.yml', () => {
  read()
  const warning = getLastCall(warn)[0]
  expect(warning).toMatchSnapshot()
})

test('returns empty object if file is not found', () => {
  const env = read({
    path: '',
  })
  expect(env).toEqual({})
})

test('throws exception if file is invalid', () => {
  expect(() => read({
    path: INVALID_CONFIG_FILE_PATH,
  })).toThrow()
})

test('environment defaults to development', () => {
  process.env.NODE_ENV = undefined
  const env = read({
    path: CONFIG_FILE_PATH
  })
  expect(env).toEqual(DEVELOPMENT_ENV)
  process.env.NODE_ENV = 'test'
})

test('environment reads from NODE_ENV', () => {
  process.env.NODE_ENV = 'development'
  const env = read({
    path: CONFIG_FILE_PATH
  })
  expect(env).toEqual(DEVELOPMENT_ENV)
  process.env.NODE_ENV = 'test'
})

test('accepts custom environment option', () => {
  const env = read({
    path: CONFIG_FILE_PATH, 
    environment: 'production'
  })
  expect(env).toEqual(PRODUCTION_ENV)
})

test('can read json file', () => {
  process.env.NODE_ENV = 'development'
  const env = read({
    path: JSON_CONFIG_FILE_PATH
  })
  expect(env).toEqual(DEVELOPMENT_ENV)
  process.env.NODE_ENV = 'test'
})