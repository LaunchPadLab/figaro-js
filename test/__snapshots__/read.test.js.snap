// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`path defaults to {project_root}/config/application.yml 1`] = `"Figaro: config file not found, skipping load (path=/Users/dpikt/Launchpad/figaro-js/config/application.yml)"`;
